File: node/packages/client/src/status.ts
/** gRPC status codes, matching @grpc/grpc-js `status`. */
export enum Status {
  OK = 0,
  CANCELLED = 1,
  UNKNOWN = 2,
  INVALID_ARGUMENT = 3,
  DEADLINE_EXCEEDED = 4,
  NOT_FOUND = 5,
  ALREADY_EXISTS = 6,
  PERMISSION_DENIED = 7,
  RESOURCE_EXHAUSTED = 8,
  FAILED_PRECONDITION = 9,
  ABORTED = 10,
  OUT_OF_RANGE = 11,
  UNIMPLEMENTED = 12,
  INTERNAL = 13,
  UNAVAILABLE = 14,
  DATA_LOSS = 15,
  UNAUTHENTICATED = 16,
}

/** A gRPC error surfaced to callers, mirroring grpc-js `ServiceError`. */
export class ServiceError extends Error {
  constructor(
    public readonly code: Status,
    public readonly details: string,
    public readonly metadata?: import("./metadata.js").Metadata,
  ) {
    super(`${Status[code] ?? code}: ${details}`);
    this.name = "ServiceError";
  }
}

File: go/webnext/status.go
package webnext

// Code is a gRPC status code. Values match the canonical gRPC set (and the codes
// the wire protocol carries in Trailer/Reset frames and Fetch trailer blocks).
type Code uint32

const (
	OK                 Code = 0
	Canceled           Code = 1
	Unknown            Code = 2
	InvalidArgument    Code = 3
	DeadlineExceeded   Code = 4
	NotFound           Code = 5
	AlreadyExists      Code = 6
	PermissionDenied   Code = 7
	ResourceExhausted  Code = 8
	FailedPrecondition Code = 9
	Aborted            Code = 10
	OutOfRange         Code = 11
	Unimplemented      Code = 12
	Internal           Code = 13
	Unavailable        Code = 14
	DataLoss           Code = 15
	Unauthenticated    Code = 16
)

// WSCloseCode maps a gRPC status to the WebSocket close code used when a stream is
// rejected pre-RPC (handshake/connection gate): 4000 + code. (See PROTOCOL.md "Auth".)
func WSCloseCode(c Code) int { return 4000 + int(c) }
